app/controllers/sessions_controller.rb

```diff
@@ -10,25 +10,45 @@ def can_authenticate?(user)
 
   def create
     @user = User.find_by(email: params.dig(:session, :email)&.downcase)
+
     if can_authenticate?(@user)
-      forwarding_url = session[:forwarding_url]
-      reset_session
-      log_in @user
-      params[:session][:remember_me] == '1' ? remember(@user) : forget(@user)
-      session[:session_token] = @user.session_token
-      redirect_to forwarding_url || @user
-    else
-      flash.now[:danger] = 'Invalid email/password combination'
-      render 'new'
+      if @user.activated?
+          forwarding_url = session[:forwarding_url]
+          reset_session
+          log_in @user
+          params[:session][:remember_me] == '1' ? remember(@user) : forget(@user)
+          session[:session_token] = @user.session_token
+          redirect_to forwarding_url || @user
+        else
+          message = "Account not activated. "
+          message += "Check your email for the activation link." 
+          flash[:warning] = message
+          redirect_to root_url
+        end
+      else
+        flash.now[:danger] = 'Invalid email/password combination'
+        render 'new'
     end
   end
-
+  # Returns the current logged-in user (if any).
   def current_user
-    if session[:user_id]
-      User.find_by(id: session[:user_id])
+    if (user_id = session[:user_id])
+      user ||= User.find_by(id: user_id)
+        @current_user||= user if (user.session_token!=session['session_token'])
```

Comment:
```suggestion
      @current_user ||= user if (user.session_token != session['session_token'])
```
I would also try to extract everything after the `if` into a method.

Although reading the implementation here, I'm a bit confused, shouldn't we set `@current_user` if the session token matches (instead of current implementation where it's set when it mismatches)?